src/lib/formats/vgi_dsk.cpp

```diff
@@ -0,0 +1,144 @@
+
+Micropolis VGI disk image format
+
+The format is essentially an HCS-ordered sector image, but with raw
+275 byte sectors. The sector format is: SYNC (0xFF), track, sector,
+10 byte "user data", 256 byte data, checksum, 4 byte ECC, ECC present
+flag.
+
+http://www.bitsavers.org/pdf/micropolis/metafloppy/1084-01_1040_1050_Users_Manual_Apr79.pdf
+
+*********************************************************************/
+
+#include <cstring>
+
+#include "vgi_dsk.h"
+#include "ioprocs.h"
+
+static const int track_size = 100'000;
+static const int half_bitcell_size = 2000;
+
+micropolis_vgi_format::micropolis_vgi_format() : floppy_image_format_t()
+{
+}
+
+struct format {
+	int head_count;
+	int track_count;
+	uint32_t variant;
+};
+
+static const format formats[] = {
+	{1, 35, floppy_image::SSDD}, // MOD-I
+	{2, 35, floppy_image::DSDD},
+	{1, 77, floppy_image::SSQD}, // MOD-II
+	{2, 77, floppy_image::DSQD},
+	{}
+};
+
+static format find_format(int file_size)
+{
+	for (int i = 0; formats[i].head_count; i++)
+		if (file_size == formats[i].head_count * formats[i].track_count * 16 * 275)
+			return formats[i];
+	return {};
+}
+
+int micropolis_vgi_format::identify(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants) const
+{
+	uint64_t file_size;
+	if (io.length(file_size))
+		return 0;
+	format fmt = find_format(file_size);
+	if (!fmt.head_count)
+		return 0;
+	if (!has_variant(variants, fmt.variant))
+		return 0;
+	return FIFID_SIZE;
+}
+
+bool micropolis_vgi_format::load(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants, floppy_image *image) const
+{
+	uint64_t file_size;
+	if (io.length(file_size))
+		return 0;
+	const format fmt = find_format(file_size);
+	if (!fmt.head_count)
+		return false;
+	image->set_variant(fmt.variant);
+
+	std::vector<uint32_t> buf;
+	uint8_t sector_bytes[275];
+	for (int head = 0; head < fmt.head_count; head++) {
+		for (int track = 0; track < fmt.track_count; track++) {
+			for (int sector = 0; sector < 16; sector++) {
+				for (int i = 0; i < 40; i++)
+					mfm_w(buf, 8, 0, half_bitcell_size);
+				std::size_t actual;
+				io.read(sector_bytes, std::size(sector_bytes), actual);
+				for (int i = 0; i < std::size(sector_bytes); i++)
+					mfm_w(buf, 8, sector_bytes[i], half_bitcell_size);
+				while (buf.size() < track_size/16 * (sector+1))
+					mfm_w(buf, 8, 0, half_bitcell_size);
+			}
+			generate_track_from_levels(track, head, buf, 0, image);
+			buf.clear();
+		}
+	}
+	return true;
+}
+
+bool micropolis_vgi_format::save(util::random_read_write &io, const std::vector<uint32_t> &variants, floppy_image *image) const
+{
+	uint32_t variant = image->get_variant();
+	format fmt = {};
+	for (int i = 0; formats[i].head_count; i++)
+		if (variant == formats[i].variant)
+			fmt = formats[i];
+	if (!fmt.head_count) {
+		int heads, tracks;
+		image->get_actual_geometry(tracks, heads);
+		printf("heads: %d tracks %d\n", heads, tracks);
```

Comment:
This `printf` shouldn’t be here – maybe use `LOG_FORMATS` instead, and make the message give a bit more context?